### PR TITLE
Crafting inventories/windows

### PR DESCRIPTION
**Pull request by kitskub**
This sets up Crafting Tables for crafting. All it needs now is work on the recipe side and searching through recipes and updating accordingly. Feel more than free to make comments/scream at how I messed up on something and the like.

Related: https://github.com/SpoutDev/SpoutAPI/pull/233'
This also partially fixes: http://issues.spout.org/browse/VANILLA-72

### CODE REVIEW

File: src/main/java/org/spout/vanilla/material/block/controlled/CraftingTableBlock.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import org.spout.api.inventory.ItemStack;
+import org.spout.api.material.block.BlockFace;
+
+import org.spout.vanilla.controller.VanillaControllerTypes;
+import org.spout.vanilla.controller.block.CraftingTable;
+import org.spout.vanilla.controller.living.player.VanillaPlayer;
+import org.spout.vanilla.inventory.CraftingTableInventory;
+import org.spout.vanilla.util.Instrument;
+
+public class CraftingTableBlock extends ControlledMaterial {
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/material/block/controlled/CraftingTableBlock.java]: Should just be "CraftingTable" the "Block" is redundant

**Comment by aPunch** [comment]
Why does CraftingTable need to be controlled? Can't recipes be checked for in CraftingTableInventory's onClick methods?

**Comment by windy1** [comment]
It doesn't and I said the same thing in IRC @Kitskub

File: src/main/java/org/spout/vanilla/material/block/controlled/CraftingTableBlock.java (hunk 1 above)
Comment (kitskub) [inline src/main/java/org/spout/vanilla/material/block/controlled/CraftingTableBlock.java]: If it's "CraftingTable" then there are a ton of errors because its the same name as the controller, so this is what zidane told me to do.

**Comment by kitskub** [comment]
Unfortunately, it does for the same reason that Enchantment table needs a controller: it holds the instance of an inventory. However, if there's another way to do this, please point it out.

**Comment by aPunch** [comment]
CraftingTable shouldn't need to hold the instance of any inventory. All it needs is an onInteractBy method that points to a new CraftingTableInventory, which will handle all crafting.

File: src/main/java/org/spout/vanilla/material/block/controlled/CraftingTableBlock.java (hunk 1 above)
Comment (LunNova) [inline src/main/java/org/spout/vanilla/material/block/controlled/CraftingTableBlock.java]: Use the full name when referencing org.spout.vanilla.controller.block.CraftingTable.

**Comment by windy1** [comment]
@aPunch is right @kitskub the enchantment table will be used for other data besides the inventory.

**Comment by kitskub** [comment]
Ok, but it actually is going to need to be a map of player to inventory, because as it is, all players see the same one.

**Comment by windy1** [comment]
No, you would create a new instance of the inventory every CraftingTable#onInteract

**Comment by kitskub** [comment]
Yeah, just looked through the code, and that'll work.

**Comment by kitskub** [comment]
Fixed

File: src/main/java/org/spout/vanilla/material/block/solid/CraftingTable.java [hunk 2]
```diff
@@ -1,54 +1,75 @@
...
+import org.spout.api.inventory.ItemStack;
+import org.spout.api.material.block.BlockFace;
+
+import org.spout.vanilla.controller.living.player.VanillaPlayer;
+import org.spout.vanilla.inventory.CraftingTableInventory;
+import org.spout.vanilla.material.block.Solid;
+import org.spout.vanilla.util.Instrument;
+
+public class CraftingTable extends Solid {
+	CraftingTableInventory inventory;
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/material/block/solid/CraftingTable.java]: No, we don't want a global field, we want to instantiate an entirely new CraftingTableInventory every CraftingTable#onInteract

(edit) s/global field/instance variable
(edit) oh.. wait... why is this here?

File: src/main/java/org/spout/vanilla/window/CraftingWindow.java [hunk 3]
```diff
@@ -26,10 +26,67 @@
...
+	
+	public CraftingWindow(VanillaPlayer owner, CraftingTableInventory craftingInventory) {
 		super(1, "Crafting", owner);
+		this.setInventory(owner.getInventory().getItems(), craftingInventory);
+		this.setSlotConversionArray(SLOTS);
 	}
+
+	@Override
+	public void onClosed() {
+		for (int slot = 0; slot < this.getInventory().getSize() - 9; slot++) {
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/window/CraftingWindow.java:45]: What is the point of this loop?

File: src/main/java/org/spout/vanilla/window/CraftingWindow.java [hunk 4]
```diff
@@ -26,10 +26,67 @@
...
+			ItemStack current = this.getInventory().getItem(getSpoutSlotIndex(slot));
+			if (current == null) {
+				continue;
+			}
+			Item control = new Item(current, getOwner().getHeadPosition().add(getOwner().getLookingAt()));
+			getOwner().getParent().getWorld().createAndSpawnEntity(getOwner().getHeadPosition().add(0.0, -0.4, 0.0), control);
+			this.getInventory().setItem(getSpoutSlotIndex(slot), null);
+		}
+        }
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/window/CraftingWindow.java]: Formatting

**Comment by windy1** [comment]
Also, slot conversion should be bottom to top and left to right.

File: src/main/java/org/spout/vanilla/material/block/solid/CraftingTable.java (hunk 2 above)
Comment (kitskub) [inline src/main/java/org/spout/vanilla/material/block/solid/CraftingTable.java]: Totally missed this when I got rid of it. Oops.

Comment (stephenwilliams) [inline src/main/java/org/spout/vanilla/material/block/solid/CraftingTable.java]: Its not global, it just needs to be change to private, right now its access level is default

File: src/main/java/org/spout/vanilla/material/block/solid/CraftingTable.java [hunk 5]
```diff
@@ -1,54 +1,75 @@
...
+	@Override
+	public void onInteractBy(Entity entity, Block block, Action action, BlockFace face) {
+		if (action == Action.RIGHT_CLICK) {
+			Controller controller = entity.getController();
+			if (!(controller instanceof VanillaPlayer)) {
+				return;
+			}
+
+			// Open the crafting table
+			(new CraftingTableInventory()).open((VanillaPlayer) controller);
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/material/block/solid/CraftingTable.java:71]: Even though this is correct I would prefer

``` java
CraftingTableInventory inventory = new CraftingTableInventory();
inventory.open((VanillaPlayer) controller);
```

File: src/main/java/org/spout/vanilla/window/CraftingWindow.java (hunk 3 above)
Comment (kitskub) [inline src/main/java/org/spout/vanilla/window/CraftingWindow.java:45]: To drop all the items when the craftingwindow is closed

File: src/main/java/org/spout/vanilla/window/CraftingWindow.java [hunk 6]
```diff
@@ -26,10 +26,67 @@
...
+	
+	@Override
+	public boolean onLeftClick(int clickedSlot) {
+		if (clickedSlot == 0) {
+			if (this.hasItemOnCursor()) {
+				return false;
+			}
+			else {
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/window/CraftingWindow.java]: Your else statement is not formatted correctly; needs to be 

``` java
if (true) {
    // something
} else {
    // something else
}
```

not

``` java
if (true) {
    // something
}
else {
    // something else
}
```

File: src/main/java/org/spout/vanilla/window/CraftingWindow.java [hunk 7]
```diff
@@ -26,10 +26,67 @@
...
+		return true;
+	}
+	
+	@Override
+	public boolean onRightClick(int clickedSlot) {
+		if (clickedSlot == 0) {
+			if (this.hasItemOnCursor()) {
+				return false;
+			}
+			else {
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/window/CraftingWindow.java]: same here

File: src/main/java/org/spout/vanilla/inventory/CraftingTableInventory.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License,
+ * the MIT license and the SpoutDev License Version 1 along with this program.
+ * If not, see <http://www.gnu.org/licenses/> for the GNU Lesser General Public
+ * License and see <http://www.spout.org/SpoutDevLicenseV1.txt> for the full license,
+ * including the MIT license.
+ */
+package org.spout.vanilla.inventory;
+
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/inventory/CraftingTableInventory.java:28]: Two lines of blank code MAX

**Comment by kitskub** [comment]
Should ALL be fixed
Except the slot stuff should be double checked as I honestly just am guessing at that.
